ext/nnstreamer/tensor_converter/tensor_converter_flexbuf.cc

```diff
@@ -0,0 +1,158 @@
+void fini_flxc (void) __attribute__ ((destructor));
+#ifdef __cplusplus
+}
+#endif /* __cplusplus */
+
+/** @brief tensor converter plugin's NNStreamerExternalConverter callback */
+static GstCaps *
+flxc_query_caps (const GstTensorsConfig *config)
+{
+  return gst_caps_from_string (GST_FLEXBUF_CAP_DEFAULT);
+}
+
+/** @brief tensor converter plugin's NNStreamerExternalConverter callback */
+static gboolean
+flxc_get_out_config (const GstCaps *in_cap, GstTensorsConfig *config)
+{
+  GstStructure *structure;
+  g_return_val_if_fail (config != NULL, FALSE);
+  gst_tensors_config_init (config);
+  g_return_val_if_fail (in_cap != NULL, FALSE);
+
+  structure = gst_caps_get_structure (in_cap, 0);
+  g_return_val_if_fail (structure != NULL, FALSE);
+
+  /* All tensor info should be updated later in chain function. */
+  config->info.info[0].type = _NNS_UINT8;
+  config->info.num_tensors = 1;
+  if (gst_tensor_parse_dimension ("1:1:1:1", config->info.info[0].dimension) == 0) {
+    ml_loge ("Failed to set initial dimension for subplugin");
+    return FALSE;
+  }
+
+  if (gst_structure_has_field (structure, "framerate")) {
+    gst_structure_get_fraction (structure, "framerate", &config->rate_n, &config->rate_d);
+  } else {
+    /* cannot get the framerate */
+    config->rate_n = 0;
+    config->rate_d = 1;
+  }
+  return TRUE;
+}
+
+/** @brief tensor converter plugin's NNStreamerExternalConverter callback
+ */
+static GstBuffer *
+flxc_convert (GstBuffer *in_buf, gsize *frame_size, guint *frames_in, GstTensorsConfig *config)
+{
+  GstBuffer *out_buf = NULL;
+  GstMemory *in_mem, *out_mem;
+  GstMapInfo in_info;
+  guint mem_size;
+  gpointer mem_data;
+
+  in_mem = gst_buffer_peek_memory (in_buf, 0);
+  if (gst_memory_map (in_mem, &in_info, GST_MAP_READ) == FALSE) {
+    ml_loge ("Cannot map input memory / tensor_converter::flexbuf.\n");
+    return NULL;
+  }
+
+  flexbuffers::Map tensors = flexbuffers::GetRoot (in_info.data, in_info.size).AsMap ();
+  config->info.num_tensors = tensors["num_tensors"].AsUInt32 ();
+
+  if (config->info.num_tensors > NNS_TENSOR_SIZE_LIMIT) {
+    nns_loge ("The number of tensors is limited to %d", NNS_TENSOR_SIZE_LIMIT);
+    goto done;
+  }
+  config->rate_n = tensors["rate_n"].AsInt32 ();
+  config->rate_d = tensors["rate_d"].AsInt32 ();
+  out_buf = gst_buffer_new ();
+  *frame_size = 0;
+  *frames_in = 1;
+
+  for (guint i = 0; i < config->info.num_tensors; i++) {
+    gchar * tensor_key = g_strdup_printf ("tensor_%d", i);
+    flexbuffers::Vector tensor = tensors[tensor_key].AsVector ();
+    config->info.info[i].name = g_strdup (tensor[0].AsString ().c_str ());
+    config->info.info[i].type = (tensor_type) tensor[1].AsInt32 ();
+
+    flexbuffers::TypedVector dim = tensor[2].AsTypedVector ();
+    for (guint j = 0; j < NNS_TENSOR_RANK_LIMIT; j++) {
+      config->info.info[i].dimension[j] = dim[j].AsInt32 ();
+    }
+    flexbuffers::Blob tensor_data = tensor[3].AsBlob ();
```

Comment:
I am not sure how FlexBuffer's AsBolb works, but 1) if it has already transferred the ownership of the data, ```g_memdup``` at line 113 would not be necessary. 2) Although it just returned the pointer of the data, you might not need to ```memdup```. I think more investigation is needed.

Reply:
`AsBlob()` simply returned the pointer of the data. IMO, memdup is required because the input buffer is unreferenced.

Reply:
Q: Is this derived from in_buf -> in_mem -> in_info? Then, is it possible that *_ref can block freeing ```in_buf``` or ```in_mem```?

Reply:
> Q: Is this derived from in_buf -> in_mem -> in_info?

Right.

>  is it possible that *_ref can block freeing in_buf or in_mem?

It does not seem appropriate to use the input GstMemory directly for output because the input and output data are different.
Only raw data from the flex buffer is included in the output buffer of the tensor converter.

Reply:
I marked it as further works.
If we can get rid of memory copy, it will be very helpful for performance.
Thanks for the suggestion.